### PR TITLE
Clean up and optimize Frodo implementation 

### PR DESCRIPTION
**Pull request by aparent**
This is a WIP so don't commit it yet.

I noticed that `#define` (from #45 ) is used for parameters in some places and in other places the params stored in the `OQS_KEX` struct are used.  It would probably be a good idea to declare all params in one place for consistency.  This would have the additional benefit of having the values known at runtime throughout the implementation.

Currently the speed up on my machine is as follows:

Before:
```
Operation                           Iterations  Total time (s)  Time(us): mean      pop. stdev    Cycles: mean      pop. stdev
LWE Frodo recommended
alice 0                                    963           1.001        1039.050          71.687         2592656          178724
bob                                        694           1.000        1441.037          78.622         3596023          195931
alice 1                                   8574           1.000         116.633          12.134          290468           30179
```

After:

```
Operation                           Iterations  Total time (s)  Time(us): mean      pop. stdev    Cycles: mean      pop. stdev
LWE Frodo recommended
alice 0                                   1140           1.001         877.693          65.233         2189929          162453
bob                                        789           1.001        1268.984          51.522         3166522          128238
alice 1                                   8563           1.000         116.789          13.649          290883           33964
```

### CODE REVIEW

File: src/kex_lwe_frodo/lwe_noise.c [hunk 1]
```diff
@@ -127,32 +104,28 @@ static int lwe_sample_n_inverse_16(uint16_t *s, const size_t n, const uint16_t *
...
-		return lwe_sample_n_inverse_12(s, n, params->cdf_table, params->cdf_table_len, rand);
+		lwe_sample_n_inverse_12(s, n, params->cdf_table, params->cdf_table_len, rand);
+		break;
 	case 16:
-		return lwe_sample_n_inverse_16(s, n, params->cdf_table, params->cdf_table_len, rand);
+		lwe_sample_n_inverse_16(s, n, params->cdf_table, params->cdf_table_len, rand);
+		break;
 	default:
-		return 0;
+		break;/* TODO: Error*/
```
Comment (dstebila) [inline src/kex_lwe_frodo/lwe_noise.c]: Could put an assert(0) in here.

**Comment by dstebila** [comment]
Cool -- great to have the improvements!  Let me know if you need anything from me.

**Comment by aparent** [comment]
Unpack looks like it currently takes up around 15% of the runtime.  It could be made much faster if it was less generic.  Not sure that is worth doing right now though.

**Comment by aparent** [comment]
Will merge this soon unless anyone sees any issues with it?

File: src/kex_lwe_frodo/lwe_macrify.c [hunk 2]
```diff
@@ -73,142 +51,92 @@ int MACRIFY(oqs_kex_lwe_frodo_mul_add_as_plus_e_on_the_fly)(uint16_t *out, const
...
-	int ret = 0;
-	uint16_t *a_cols = NULL;
-	uint16_t *a_cols_t = NULL;
-
-	for (i = 0; i < PARAMS_NBAR; i++) {
-		for (j = 0; j < PARAMS_N; j++) {
-			out[i * PARAMS_N + j] = e[i * PARAMS_N + j];
-		}
-	}
-
```
Comment (aparent) [inline src/kex_lwe_frodo/lwe_macrify.c]: e is copied into on line 87:

`memcpy(out, e, PARAMS_NBAR * PARAMS_N * sizeof(uint16_t));`

File: src/kex_lwe_frodo/lwe_macrify.c [hunk 3]
```diff
@@ -11,60 +10,39 @@ void MACRIFY(oqs_kex_lwe_frodo_round2)(unsigned char *out, uint16_t *in) {
...
-	int ret = 0;
-	uint16_t *a_row = NULL;
-	uint16_t *s_transpose = NULL;
-
-	for (i = 0; i < PARAMS_N; i++) {
-		for (j = 0; j < PARAMS_NBAR; j++) {
-			out[i * PARAMS_NBAR + j] = e[i * PARAMS_NBAR + j];
-		}
-	}
-
```
Comment (aparent) [inline src/kex_lwe_frodo/lwe_macrify.c]: e is copied as mentioned below.

File: src/kex_lwe_frodo/lwe_macrify.c (hunk 2 above)
Comment (dstebila) [inline src/kex_lwe_frodo/lwe_macrify.c]: I see that now.

**Comment by aparent** [comment]
Is it ok now with these changes?

**Pull request merged**